dhall/src/Dhall/TH.hs

```diff
@@ -179,26 +187,53 @@ toNestedHaskellType haskellTypes = loop
             predicate haskellType =
                 Core.judgmentallyEqual (code haskellType) dhallType
 
-derivingClauses :: [DerivClause]
-derivingClauses =
-    [ DerivClause (Just StockStrategy) [ ConT ''Generic ]
-    , DerivClause (Just AnyclassStrategy) [ ConT ''FromDhall, ConT ''ToDhall ]
-    ]
+-- | A deriving clause for `Generic`.
+derivingGenericClause :: DerivClause
+derivingGenericClause = DerivClause (Just StockStrategy) [ ConT ''Generic ]
+
+-- | Generates a `FromDhall` instances.
+fromDhallInstance
+    :: Syntax.Name -- ^ The name of the type the instances is for
+    -> Q Exp       -- ^ A TH splice generating some `Dhall.InterpretOptions`
+    -> Q [Dec]
+fromDhallInstance n interpretOptions = [d|
+    instance FromDhall $(pure $ ConT n) where
+        autoWith = Dhall.genericAutoWithInputNormalizer $(interpretOptions)
+    |]
+
+-- | Generates a `ToDhall` instances.
+toDhallInstance
+    :: Syntax.Name -- ^ The name of the type the instances is for
+    -> Q Exp       -- ^ A TH splice generating some `Dhall.InterpretOptions`
+    -> Q [Dec]
+toDhallInstance n interpretOptions = [d|
+    instance ToDhall $(pure $ ConT n) where
+        injectWith = Dhall.genericToDhallWithInputNormalizer $(interpretOptions)
+    |]
 
 -- | Convert a Dhall type to the corresponding Haskell datatype declaration
 toDeclaration
     :: (Eq a, Pretty a)
-    => [HaskellType (Expr s a)]
+    => GenerateOptions
+    -> [HaskellType (Expr s a)]
     -> HaskellType (Expr s a)
-    -> Q Dec
-toDeclaration haskellTypes MultipleConstructors{..} =
+    -> Q [Dec]
+toDeclaration generateOptions@GenerateOptions{..} haskellTypes typ@MultipleConstructors{..} =
     case code of
         Union kts -> do
             let name = Syntax.mkName (Text.unpack typeName)
 
-            constructors <- traverse (toConstructor haskellTypes typeName) (Dhall.Map.toList kts )
+            let derivingClauses =
+                    [ derivingGenericClause | generateFromDhallInstance || generateToDhallInstance ]
 
-            return (DataD [] name [] Nothing constructors derivingClauses)
+            constructors <- traverse (toConstructor generateOptions haskellTypes typeName) (Dhall.Map.toList kts)
+
+            let interpretOptions = generateToInterpretOptions generateOptions typ
+
+            fmap concat . sequence $
```

Comment:
I noticed that there is a `Monoid Q` instance but I decided not to use it since it requires `template-haskell-2.17.0.0`.
As an experiment, I set `template-haskell >= 2.17.0.0 && < 2.18` in the `dhall` Cabal file and I went straight to Dependency Hell.
Besides that, even if it was working I didn't want to bump the lower bounds just for that since it would require GHC > 9 (or building `template-haskell >= 2.17.0.0` separately). And I am not sure if it is a bit to early for that.

Reply:
Ah, never mind, then.  I didn't realize it wasn't available in older `template-haskell` versions